### ISSUE TITLE
SuperfluousWhitespaceSniff does not find trailing whitespace in blank lines

### ISSUE THREAD

**Issue by dereuromark** [squizlabs/PHP_CodeSniffer#2058]
SuperfluousWhitespaceSniff does not find the following trailing whitespace in a blank line:

```php
    public function __construct(CompanyBusinessUnitRepositoryInterface $repository)
    {
        $this->repository = $repository;
    }
    [4 whitespace space characters here and not fixed!]
    /**
     * @param \Generated\Shared\Transfer\CompanyUserResponseTransfer $companyUserResponseTransfer
     *
     * @return \Generated\Shared\Transfer\CompanyUserResponseTransfer
     */
    public function assignDefaultBusinessUnitToCompanyUser(
```

I guess the whitespace here cannot be found/fixed due to the following doc block (and as such not a functional character)? Probably requires a better strategy.
```php
    }
    
    /**
```

**Comment by gsherwood** [comment]
When I check your sample code, I get `Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)` for that line you indicated, which is also fixed correctly.

The whitespace shouldn't have anything to do with the docblock:
```diff
--- temp.php
+++ PHP_CodeSniffer
@@ -3,7 +3,7 @@
 {
     $this->repository = $repository;
 }
-
+
 /**
  * @param \Generated\Shared\Transfer\CompanyUserResponseTransfer $companyUserResponseTransfer
  *
```
Are you sure you are not muting this specific error in your standard or another you are importing? Are you able to throw some sample code into a gist so I can copy/paste to replicate?

**Comment by dereuromark** [comment]
Pretty sure https://github.com/spryker/code-sniffer/blob/master/Spryker/ruleset.xml#L119 includes and runs this normally.

**Comment by gsherwood** [comment]
Can you replicate with `phpcs --standard=Squiz --sniffs=Squiz.WhiteSpace.SuperfluousWhitespace /path/to/file.php`? If so, I really need some content to help replicate the problem.

**Comment by dereuromark** [comment]
Some debugging reveals that ignoreBlankLines is set from false to true here.
How does that happen suddenly? I don't overwrite it, and the default is false.

**Comment by dereuromark** [comment]
OK, turns out that if you include PSR2 (those knuckleheads...), you suddenly get all whitespace lines dropped...

Readding this fixes things:
```xml
    <rule ref="Squiz.WhiteSpace.SuperfluousWhitespace">
        <properties>
            <property name="ignoreBlankLines" value="false"/>
        </properties>
    </rule>
```

I highly recommend stopping this very bad idea of just disabling this fixer, and instead adjust psr2 properly by using proper begin and end of construct whitespace removers and letting this one do only what it is intended to do: remove trailing whitespace in newlines.

**Comment by gsherwood** [comment]
> I highly recommend stopping this very bad idea of just disabling this fixer, and instead adjust psr2 properly by using proper begin and end of construct whitespace removers and letting this one do only what it is intended to do: remove trailing whitespace in newlines.

I think you're suggesting that I write additional sniffs to find and fix independent types of whitespace. Although "end of construct whitespace" isn't a thing - it's whitespace at the end of lines and whitespace on empty lines that differs. So I'd need a sniff that detects and removes whitespace at the end of non-empty lines and one that does the exact same thing except only for empty lines.

That's certainly possible, but feels like a lot of repeated code.

If you think 2 new sniffs to check for whitespace is worth doing, I'd be happy to review a PR.

And alternative might be to provide a clearer way of seeing what a standard you are including is doing with sniff settings. You can see these using `-vv` already (check the ruleset parsing output at the top) but maybe there is another way of presenting this information to help figure out these sort of things.

Either way, I'm going to close this issue now that you've figured it out (thanks for sharing the solution).

**Comment by dereuromark** [comment]
Yes, two independent sniffs would be best and most clear.
